server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ChunkOutputStream.java

```diff
@@ -1,156 +1,147 @@
+    ) {
+        this.completionSink = Objects.requireNonNull(completionSink);
+        this.parentResponse = Objects.requireNonNull(parentResponse);
+        this.nettyOutbound = Objects.requireNonNull(reactorNettyResponse);
+    }
+
+    @Override
+    public void write(int b) {
+        write(new byte[]{(byte) b}, 0, 1);
+    }
+
+    @Override
+    public void write(byte[] bs, int off, int len) {
+        try {
+            asyncWrite(bs, off, len).toCompletableFuture().get();
+        } catch (final InterruptedException ie) {
+            Thread.currentThread().interrupt();
+            throw new ReactorNettySendException(ie);
+        } catch (final ExecutionException ee) {
+            throw new ReactorNettySendException(ee);
         }
-
-        byte[] bytes = bs;
-        if (offset != 0 || length != bs.length) {
-            bytes = Arrays.copyOfRange(bs, offset, offset + length);
+    }
+
+    @Override
+    public void flush() {
+        try {
+            asyncFlush().get();
+        } catch (final InterruptedException ie) {
+            Thread.currentThread().interrupt();
+            throw new ReactorNettySendException(ie);
+        } catch (final ExecutionException ee) {
+            throw new ReactorNettySendException(ee);
+        }
+    }
+
+    @Override
+    public CompletableFuture<Void> asyncFlush() {
+        // Everything flows through asyncWrite and we are flushing on each call, so we
+        // will treat that as a no-op for now, assuming that callers of this would have
+        // chained this onto an asyncWrite.  I hope that doesn't mess up SSE..
+        return CompletableFuture.completedFuture(null);
+    }
+
+    static class ReactorNettySendException extends RuntimeException {
+        ReactorNettySendException(final Throwable cause) {
+            super(cause);
         }
-        byteSink.emitNext(Tuples.of(bytes, cf), EMIT_FAILURE_HANDLER);
-        return cf;
-   }
+    }
+
+    @Override
+    public CompletableFuture<Void> asyncWrite(final byte[] bs, int offset, int length) {
+        try {
+            if (!started) {
+                parentResponse.committed();
+                started = true;
+            }
+            byte[] bytes = bs;
+            if (offset != 0 || length != bs.length) {
+                bytes = Arrays.copyOfRange(bs, offset, offset + length);
+            }
+
+            return nettyOutbound
+                    .send(Mono.just(Unpooled.wrappedBuffer(bytes)), FLUSH_ON_EACH_WRITE)
+                    .then()
+                    .doOnError(err -> completionSink.emitError(err, Sinks.EmitFailureHandler.FAIL_FAST))
+                    .doOnCancel(() -> completionSink.emitError(
+                            RESPONSE_WRITE_ABORTED_ON_CANCEL,
+                            Sinks.EmitFailureHandler.FAIL_FAST
+                    ))
+                    .doOnDiscard(
+                            ByteBuf.class,
+                            byteBuf -> completionSink.emitError(
```

Comment:
Could you use netty technique for detecting mem leak and see if any happens?  You may need to call `ByteBuf#release` here, but I kind of doubt it..

Reply:
I can see that the ByteBuf release on discard is handled in NettyOutbound#send [here](https://github.com/reactor/reactor-netty/blob/main/reactor-netty-core/src/main/java/reactor/netty/channel/ChannelOperations.java#L287).

Reply:
The reason I think that's an acceptable answer is because of [this](https://netty.io/wiki/reference-counted-objects.html#who-destroys-it).  Otherwise, I'd say we can't rely on other's implementation (unless we must).